[CEP] commcare extensions

snopoke opened the pull request:
##### SUMMARY
Potential extension system for CommCare

Consider this a rough draft to get feedback.

snopoke committed:
commcare plugins

czue commented on corehq/plugins/signal_plugin.py:
```diff
@@ -0,0 +1,41 @@
+class SignalPlugins(Plugins):
+    """Plugin implementation that uses Django signals to manage the
+    dispatching of extension point events
+    """
```
Are you anticipating there would be other implementations of `Plugins`? What would be the use case for that?

snopoke replied:
not necessarily but it does make it easier if we do

czue reviewed (commented):
This is neat. I don't have a lot of experience with plug-in system designs, but the plugin-facing API as well as the hook-creation process both look quite clean which I think is probably the most important part. Did you model it off any existing system?

It looks like plugins are intentionally managed at the environment level. Was there any consideration to do anything at the domain level? I realize the plugin itself can - and likely will - manage that, but it might be nice to have some explicit way of saying "enable this plug in for this domain". I'm just picturing a lot of our initial use cases will be extracting custom reports code, and it might be nice to centralize the logic of where that code runs for an environment like prod.

snopoke commented:
another option for hooking things up would be something like this:

```python
PLUGINS = {
    "pointa": {
        "callable": "a",
        "domains": [],
    }
}
```

re the domain enablement, I had envisioned that check happening in the plugin itself but it could also be part of the framework (as in the example above).

One advantage of the settings approach is that it's easy to see what the full list of plugins is.

snopoke committed:
manage plugins via settings and without signals

snopoke commented:
Also found this https://pluggy.readthedocs.io/en/latest/

millerdev commented:
> Also found this https://pluggy.readthedocs.io/en/latest/

Pluggy looks interesting, and I'm definitely +1 on using a an existing plugin system rather than inventing our own if it involves sufficient complexity.

I didn't take a lot of time to evaluate Pluggy yet, but the fact that it's the pytest plugin system seems both good and bad to me. I'm not sure if Pluggy implements the thing used for pytest fixtures, but I find them a bit too magicical. I'm thinking specifically of the feature that automagically maps global names to test method arguments. Is that behavior enabled by Pluggy? I'd prefer not to introduce that kind of magic into the HQ code base.

snopoke commented:
Doing a quick look at pluggy again I think it would work well but one thing we couldn't do is filter which plugins to call by domain. That would push the domain checking down to the plugin instead of the framework. Not sure how important that is. 

One other thing is that any exception in a plugin prevents other plugins from being called and bubbles up to the caller which doesn't seem ideal for our use case.

snopoke committed:
new extension point for icds urls

snopoke committed:
better exceptions

snopoke committed:
move extension point to parent class

snopoke commented on urls.py:
```diff
@@ -92,6 +90,11 @@
 
+for plugin_url_modules in plugins.get_contributions("urls:domain_specific"):
+    for module in plugin_url_modules:
+        domain_specific.append(url(r'^', include(module)))
```
the alternative is to put these in locasettings: https://github.com/dimagi/commcare-hq/pull/28094/files#diff-de6dd4b4c889fe0882cfd3f6df5aa451R92

millerdev commented on corehq/plugins/interface.py:
```diff
@@ -11,36 +14,77 @@ class ExtensionPoint:
+        if isinstance(_callable, str):
+            _callable = to_function(self.callable_ref)
+        if not _callable or not callable(_callable):
+            raise Exception(f"Plugin callable is not callable: '{self.callable_ref}'")
```
This module raises `Exception` in several places, which cannot safely be caught without possibly catching other unwanted exceptions. What do you think of adding a custom exception type to be used by this module?

Could be as simple as adding this at the bottom of this module, and then `raise Error(...)`
```py
class Error(Exception):
    pass
```

millerdev commented on corehq/plugins/interface.py:
```diff
@@ -0,0 +1,92 @@
+        return f"{self.callable_ref}"
+
+
+class Plugins:
```
Non-blocking: would it make sense to call these things "Extensions" instead of "Plugins" since they consist of "extension points"? Similar to how Firefox and Chrome have extensions (Java and Flash were plugins, but they are being phased out).

snopoke committed:
use custom exception + fix tests

snopoke committed:
rename

snopoke committed:
refactor register

snopoke commented on corehq/extensions/__init__.py:
```diff
@@ -0,0 +1,7 @@
+from corehq.extensions.interface import CommCareExtensions, ExtensionError
```
```suggestion
from corehq.extensions.interface import CommCareExtensions, ExtensionError   # noqa
```

snopoke commented on corehq/extensions/tests.py:
```diff
@@ -0,0 +1,92 @@
+
+    with testil.assert_raises(ExtensionError, msg=re.compile("consumed.*arg1")):
+        extensions.register_extension("ext_point_a", bad_spec)
+
```
```suggestion
```

snopoke committed:
lint

millerdev commented on corehq/extensions/__init__.py:
```diff
@@ -1,4 +1,4 @@
-from corehq.extensions.interface import CommCareExtensions, ExtensionError
+from corehq.extensions.interface import CommCareExtensions, ExtensionError   # noqa
```
Typically `noqa` is followed by a specific error or warning code to be ignored. Is there a reason why you're completely turning off flake8 on lines like this in this PR?

snopoke committed:
rename app config

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

snopoke committed:
remove reference to plugin

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

snopoke commented on corehq/extensions/app_config.py:
```diff
@@ -0,0 +1,10 @@
+    name = 'corehq.extensions'
+
+    def ready(self):
+        from corehq.extensions import extension_points, extension_manager  # noqa
```
so that the extension points get registered:

```
register_extension_point(name, ..etc)
```

millerdev replied:
I thought that was another object from the same module, but I see now it's a different module.

snopoke commented:
I am thinking of changing how the extension points are defined to make it simpler to include them in the sphinx docs:

```python
@extension_point
def my_extension_point(arg1, arg2):
    """My extension point docs"""

@extension_implementation
def my_extension_point(arg1, arg2):
     # do something
```

millerdev commented:
The only issue I see here is the binding of extension point to implementation seems a bit tenuous since they are only linked by having the same name. How about something like this:
```py
@extension_point
def my_extension_point(arg1, arg2):
    """My extension point docs"""

@my_extension_point.implementation
def my_extension_point_name_doesnt_matter(arg1, arg2):
     # do something
```
Hopefully uses of the phrase "extension point" will not be so dense in reality. What do you think of this:
```py
@extension_point
def domain_specific_urls():
    """Return a list domain specific of URL configs

    Each returned item is a dict consisting of a type specifier (key)
    and a module string (value). Valid type specifiers are:

    - "callable"
    - "..."
    TODO improve/complete this documentation

    Module strings will be included in the domain specific URL patterns.
    """

@domain_specific_urls.extend
def icds_urls():
    return [
        {"callable": "custom.icds.commcare_extensions.urls_domain_specific"}
    ]
```

snopoke committed:
correct app name

snopoke committed:
remove unused import

snopoke commented:
@millerdev I do like that option. It does force you to import the extension point function as well which may add a bit of safely

snopoke committed:
convert to using decorators to define extensions

snopoke committed:
lint

millerdev commented on corehq/extensions/interface.py:
```diff
@@ -0,0 +1,146 @@
+    """Decorator for creating an extension point"""
+    def extend(domains=None):
+
+        def _extend(impl):
```
Would you agree it would be useful to have this raise an error if it is called after `corehq.extensions.app_config.ExtensionAppConfig.ready()`?

This could also add the extension to a list (something like `ExtensionAppConfig.extensions`), which would make it possible to eliminate the module traversal currently used to find extensions.

snopoke replied:
I made an attempt at this here: fcd751a36cb but it's a bit messy

millerdev replied:
I reworked it a bit to do what I had in mind. Feel free to cherry-pick https://github.com/dimagi/commcare-hq/commit/f623e8a1f5fc47027f65360154811e0947ee5d17 onto your branch if you like it (also feel free to modify it if you see issues).

millerdev commented on urls.py:
```diff
@@ -92,6 +90,11 @@
     url(r'^submit_feedback/$', submit_feedback, name='submit_feedback'),
 ]
 
+for url_modules in extensions.get_contributions("urls:domain_specific"):
```
Are there tests for this? I would have expected it to raise an error.
```suggestion
for url_modules in extensions.get_contributions("domain_specific_urls"):
```

millerdev commented on corehq/tabs/uitab.py:
```diff
@@ -90,11 +91,20 @@ def dropdown_items(self):
+        items.extend([
+            dropdown_dict(**response)
+            for response in extensions.get_contributions(
+                "uitab:dropdown_items", tab=tab_name, domain=self.domain, request=self._request
```
```suggestion
                "uitab_dropdown_items", tab=tab_name, domain=self.domain, request=self._request
```

snopoke commented on corehq/extensions/interface.py:
```diff
@@ -15,7 +16,24 @@ class ExtensionError(Exception):
+
+def extension_point(func):
+    """Decorator for creating an extension point"""
+    def extend(domains=None):
+
+        def _extend(impl):
+            setattr(impl, "extension_point_impl", {
+                "spec": func.__name__,
+                "domains": domains
+            })
+            return impl
+
+        return _extend
```
how about this:
```suggestion
    def extend(domains=None):

        def _extend(impl):
            setattr(impl, "extension_point_impl", {
                "spec": func.__name__,
                "domains": domains
            })
            return impl

        return _extend
```

```python
def extend(domains=None);
```suggestion
    def extend(domains=None):

        def _extend(impl):
            setattr(impl, "extension_point_impl", {
                "spec": func.__name__,
                "domains": domains
            })
            return impl

        # if `domains` is callable it means the decorator was used without being called first
        return _extend(impl) if callable(domains) else _extend
```

snopoke replied:
Took me a while to figure out what the `*,` was doing

snopoke committed:
pr feedback

snopoke committed:
update extension_point decorator to be more flexible + docs

snopoke committed:
Merge branch 'master' into sk/plugins

snopoke committed:
prevent late extension definition

millerdev committed:
Avoid module traversal

Extension points and extensions are now scoped to the extension manager with which they are defined, which means less globals and allows cleaner testing.

snopoke committed:
remove unneeded call to add_extension_points

snopoke commented:
@millerdev there's one other thing I was thinking to change which is how the callers get the results from the extension points. Currently the pattern is this:
```python
extensions.get_contributions("ext_point_name", *args, **kwargs)

# I was thinking to change that to a normal function call:
extensions.hooks.ext_point_name(*args, **kwargs)
```

Let me know what you think.

millerdev commented on corehq/extensions/interface.py:
```diff
@@ -0,0 +1,154 @@
+    def __init__(self):
+        self.registry = defaultdict(list)
+        self.extension_point_registry = {}
+        self.extensions = []
```
```suggestion
        self.locked = False
```

millerdev commented on corehq/extensions/interface.py:
```diff
@@ -0,0 +1,154 @@
+                    f"Incorrect usage of extension decorator. See docs below:"
+                    f"\n\n{self.extension_point.__doc__}"
+                )
+                self.extensions.append(Extension(func.__name__, impl, domains))
```
```suggestion
                self.register_extension(Extension(func.__name__, impl, domains))
```

millerdev commented on corehq/extensions/interface.py:
```diff
@@ -0,0 +1,154 @@
+        func.extend = extend
+        return func
+
+    @property
+    def locked(self):
+        return self.extensions is None
+
```
```suggestion
```

millerdev commented on corehq/extensions/interface.py:
```diff
@@ -0,0 +1,154 @@
+    def load_extensions(self, implementations):
+        for module_name in implementations:
+            self.resolve_module(module_name)
+        for extension in self.extensions:
+            self.register_extension(extension)
+        self.extensions = None
```
```suggestion
        self.locked = True
```

millerdev commented:
> ```py
> extensions.hooks.ext_point_name(*args, **kwargs)
> ```
>
> Let me know what you think.

:+1: I like that.

snopoke committed:
Register extensions directly

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

snopoke committed:
make extensions calls function calls

snopoke committed:
add docs

snopoke committed:
make extension points the callable to return results

snopoke commented:
@millerdev I initially did what I had suggested (a2081cba193)

Then I realized that we could make the actual extension point functions into the hook caller which I did in: a2081cba193. I think this is the cleanest implementation makes it easier to spot mistakes in names etc since it appears as just a normal function call so any static analysis etc will work as expected.

```python
@extension_point
def get_things():
    pass

results = get_things()
```

snopoke committed:
imports

snopoke committed:
remove bad import

snopoke committed:
add extensions doc root

millerdev reviewed (approved):
> https://github.com/dimagi/commcare-hq/pull/28050/commits/a2081cba193f4df2ff71f92844eae34bce4c6199 make extension points the callable to return results

This is a great idea! Why didn't I think of that? :stuck_out_tongue:

snopoke commented on corehq/extensions/interface.py:
```diff
@@ -44,9 +44,43 @@ def __repr__(self):
+        results = []
+        for extension in self.extensions:
+            try:
+                result = extension(**kwargs)
```
hmm, clearly

snopoke committed:
Apply suggestions from code review

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

snopoke committed:
remove old method of adding custom urls

snopoke committed:
more docs

snopoke committed:
Merge branch 'sk/plugins' of github.com:dimagi/commcare-hq into sk/plugins

snopoke committed:
pass args all the way through + better domain arg support

snopoke committed:
check if domain filtering is supported

snopoke committed:
update docs

snopoke committed:
typo

snopoke committed:
remove _Registry object

snopoke committed:
command to list extension points

snopoke committed:
return a list

snopoke commented on docs/extensions.rst:
```diff
@@ -0,0 +1,10 @@
+.. automodule:: corehq.extensions
+
+Extension points
```
I tried to use autodocs to add docs for the extension points but it didn't work well (presumably because the functions are replaced by the callables).

snopoke commented:
@millerdev I've done a bit of tidy up and documentation. I think this is GTG

snopoke committed:
add management command

millerdev commented on corehq/extensions/interface.py:
```diff
@@ -0,0 +1,113 @@
+        if unconsumed_args and not spec.varkw:
+            raise ExtensionError(f"Not all extension point args are consumed: {unconsumed_args}")
+
+    def call_for_domain(self, domain):
```
This does not call anything. I think the intent was clearer with a `should_` prefix.
```suggestion
    def should_call_for_domain(self, domain):
```

millerdev commented:
I like how this turned out.

snopoke committed:
Update corehq/extensions/interface.py

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

snopoke commented on corehq/extensions/interface.py:
```diff
@@ -0,0 +1,113 @@
+        callargs = inspect.getcallargs(self.definition_function, *args, **kwargs)
+        domain = callargs.get('domain')
+        for extension in self.extensions:
+            if domain and not extension.call_for_domain(domain):
```
```suggestion
            if domain and not extension.should_call_for_domain(domain):
```

snopoke committed:
Update corehq/extensions/interface.py

snopoke committed:
Merge branch 'master' into sk/plugins

snopoke committed:
remove old setting

snopoke committed:
selectively add commcare extensions

snopoke committed:
update readme
